nixos/borgmatic: added test

flandweber opened the pull request:
Added validation of borgmatic config to system.checks to test it at build time.
Added basic nixos test to ensure functionality.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

flandweber commented:
@ofborg test borgmatic

imlonghao commented on nixos/modules/services/backup/borgmatic.nix:
```diff
@@ -100,5 +100,10 @@ in
     systemd.timers.borgmatic.wantedBy = [ "timers.target" ];
+
+    system.checks = singleton (pkgs.runCommandCC "config validation" { } ''
+        ${pkgs.borgmatic}/bin/borgmatic -c ${cfgfile} config validate
```
`cfgfile` could be empty if using another type of format, see line L92-L97

Can `system.checks` inherit `systemd.services.borgmatic.serviceConfig.EnvironmentFile`? Because I save the `encryption_passphrase` in environment.

flandweber replied:
> cfgfile could be empty if using another type of format, see line L92-L97

good point, thanks!

> Can system.checks inherit systemd.services.borgmatic.serviceConfig.EnvironmentFile? Because I save the encryption_passphrase in environment.

system.checks is executed at build time, which means that its not possible without loading the file to the nix store.
I'm not sure if a missing `encryption_passphrase` would fail the check though.
Do you think it'd make sense to add an option to disable the check like with [varnish](https://github.com/NixOS/nixpkgs/blob/c5101e457206dd437330d283d6626944e28794b3/nixos/modules/services/web-servers/varnish/default.nix#L16)?

imlonghao replied:
> I'm not sure if a missing encryption_passphrase would fail the check though.

It will fail with exit 1

```
# head -2 /etc/borgmatic.d/filebrowser.yaml
compression: zstd
encryption_passphrase: ${FILEBROWSER_BORG_PASSPHRASE}

# borgmatic config validate
/etc/borgmatic.d/filebrowser.yaml: Error parsing configuration file
Cannot find variable FILEBROWSER_BORG_PASSPHRASE in environment
```

> Do you think it'd make sense to add an option to disable the check like with varnish?

Sounds good

flandweber committed:
nixos/borgmatic: added test

flandweber committed:
nixos/borgmatic: refactor

added configuration check at built time
added borgmatic.enableConfigCheck whether or not it'll be executed.

imlonghao reviewed (approved):
LGTM

gador reviewed (approved):
LGTM. Tested locally in a VM
